File: leetcode/easy/20_valid_parentheses/valid_parentheses.py
class Solution:
    pair = {'{': '}', '(': ')', '[': ']'}

    def isValid(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False
        stack = []
        for char in s:
            # if the next character is an open bracket it is accepted
            if char in ('(', '{', '['):
                stack.append(char)
            # if the stack is empty and the next character is a close
            # bracket then we know it cannot match
            elif len(stack) == 0:
                return False
            # if we have a close bracket and the previous character is not its
            # matching open bracket then it is invalid
            elif self.pair[stack.pop()] != char:
                return False
        # if we finish parsing and there are still characters on the stack it
        # means that more brackets were opened than closed
        if len(stack) > 0:
            return False
        return True

File: leetcode/easy/20_valid_parentheses/valid_parentheses.cpp
#include <string>
#include <vector>

class Solution {
public:
  bool isValid(std::string s) {
    std::vector<char> stack;
    auto curr = 0;
    while (curr < s.length()) {
      if (s[curr] == '(' || s[curr] == '{' || s[curr] == '[') {
        stack.push_back(s[curr]);
        curr++;
      } else if (stack.size() > 0 &&
                 (stack[stack.size() - 1] == '(' && s[curr] == ')' ||
                  stack[stack.size() - 1] == '[' && s[curr] == ']' ||
                  stack[stack.size() - 1] == '{' && s[curr] == '}')) {
        curr++;
        stack.pop_back();
      } else {
        return false;
      }
    }
    return stack.size() == 0;
  }
};
